kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java

```diff
@@ -0,0 +1,291 @@
+
+        @Override
+        public void process() {
+            if (!coordinator.isActive()) {
+                return;
+            }
+
+            try {
+                List<String> topicsDeletions = metadataStore.getChildren(getDeleteTopicsPath()).join();
+
+                HashSet<String> topicsFullNameDeletionsSets = Sets.newHashSet();
+                HashSet<KopTopic> kopTopicsSet = Sets.newHashSet();
+                topicsDeletions.forEach(topic -> {
+                    KopTopic kopTopic = new KopTopic(topic);
+                    kopTopicsSet.add(kopTopic);
+                    topicsFullNameDeletionsSets.add(kopTopic.getFullName());
+                });
+
+                log.debug("Delete topics listener fired for topics {} to be deleted", topicsDeletions);
+                Iterable<GroupMetadata> groupMetadataIterable = coordinator.getGroupManager().currentGroups();
+                HashSet<TopicPartition> topicPartitionsToBeDeletions = Sets.newHashSet();
+
+                groupMetadataIterable.forEach(groupMetadata -> {
+                    topicPartitionsToBeDeletions.addAll(
+                            groupMetadata.collectPartitionsWithTopics(topicsFullNameDeletionsSets));
+                });
+
+                Set<String> deletedTopics = Sets.newHashSet();
+                if (!topicPartitionsToBeDeletions.isEmpty()) {
+                    coordinator.handleDeletedPartitions(topicPartitionsToBeDeletions);
+                    Set<String> collectDeleteTopics = topicPartitionsToBeDeletions
+                            .stream()
+                            .map(TopicPartition::topic)
+                            .collect(Collectors.toSet());
+
+                    deletedTopics = kopTopicsSet.stream().filter(
+                            kopTopic -> collectDeleteTopics.contains(kopTopic.getFullName())
+                    ).map(KopTopic::getOriginalName).collect(Collectors.toSet());
+
+                    deletedTopics.forEach(deletedTopic -> {
+                        metadataStore.delete(
+                                getDeleteTopicsPath() + "/" + deletedTopic, Optional.of((long) -1));
+                    });
+                }
+
+                log.info("GroupMetadata delete topics {}, no matching topics {}",
+                        deletedTopics, Sets.difference(topicsFullNameDeletionsSets, deletedTopics));
+
+            } catch (Exception e) {
+                log.error("DeleteTopicsEvent process have an error {}", e.getMessage());
+                e.printStackTrace();
+            }
+        }
+    }
+
+    class BrokersChangeEvent implements KopEvent {
+        @Override
+        public void process() {
+            try {
+                getBrokers(metadataStore.getChildren(getBrokersChangePath()).join());
+            } catch (Exception e) {
+                log.error("BrokersChangeEvent process have an error {}", e.getMessage());
+                e.printStackTrace();
+            }
```

Comment:
Synchronous call in callback is easily to cause deadlock problem. Could you change to asynchronous calls? For example,

```java
            metadataStore.getChildren(getBrokersChangePath()).whenComplete((brokers, e) -> {
                if (e != null) {
                    log.warn("BrokersChangeEvent process have an error {}", e.getMessage());
                    return;
                }
                getBrokers(brokers); // maybe it should also be async
            });
```

Or is there any reason to use the synchronous call?

Reply:
good, you are right. I have no objection.